src/java/org/apache/cassandra/index/sai/memory/RowMapping.java

```diff
@@ -0,0 +1,155 @@
+    };
+
+    private final InMemoryTrie<Long> rowMapping = new InMemoryTrie<>(BufferType.OFF_HEAP);
+
+    private volatile boolean complete = false;
+
+    public PrimaryKey minKey;
+    public PrimaryKey maxKey;
+
+    public long maxSegmentRowId = -1;
+
+    private RowMapping()
+    {}
+
+    /**
+     * Create row mapping for FLUSH operation only.
+     */
+    public static RowMapping create(OperationType opType)
+    {
+        if (opType == OperationType.FLUSH)
+            return new RowMapping();
+        return DUMMY;
+    }
+
+    /**
+     * Merge IndexMemtable(index term to PrimaryKeys mappings) with row mapping of a sstable
+     * (PrimaryKey to RowId mappings).
+     *
+     * @param index a Memtable-attached column index
+     *
+     * @return iterator of index term to postings mapping exists in the sstable
+     */
+    public Iterator<Pair<ByteComparable, LongArrayList>> merge(MemtableIndex index)
+    {
+        assert complete : "RowMapping is not built.";
+
+        Iterator<Pair<ByteComparable, PrimaryKeys>> iterator = index.iterator();
+        return new AbstractGuavaIterator<Pair<ByteComparable, LongArrayList>>()
+        {
+            @Override
+            protected Pair<ByteComparable, LongArrayList> computeNext()
+            {
+                while (iterator.hasNext())
+                {
+                    Pair<ByteComparable, PrimaryKeys> pair = iterator.next();
+
+                    LongArrayList postings = null;
+                    Iterator<PrimaryKey> primaryKeys = pair.right.iterator();
+
+                    while (primaryKeys.hasNext())
+                    {
+                        PrimaryKey primaryKey = primaryKeys.next();
+                        ByteComparable byteComparable = primaryKey::asComparableBytes;
+                        Long segmentRowId = rowMapping.get(byteComparable);
+
+                        if (segmentRowId != null)
+                        {
+                            postings = postings == null ? new LongArrayList() : postings;
+                            postings.add(segmentRowId);
+                        }
+                    }
+                    if (postings != null && !postings.isEmpty())
+                        return Pair.create(pair.left, postings);
+                }
+                return endOfData();
+            }
+        };
+    }
+
+    /**
+     * Complete building in memory RowMapping, mark it as immutable.
+     */
+    public void complete()
+    {
+        assert !complete : "RowMapping can only be built once.";
+        this.complete = true;
+    }
+
+    /**
+     * Include PrimaryKey to RowId mapping
+     */
+    public void add(PrimaryKey key, long sstableRowId) throws InMemoryTrie.SpaceExhaustedException
+    {
+        assert !complete : "Cannot modify built RowMapping.";
+
+        ByteComparable byteComparable = key::asComparableBytes;
+        rowMapping.apply(Trie.singleton(byteComparable, sstableRowId), (existing, neww) -> neww);
```

Comment:
nit: I'm guessing the compiler will try to be smart about `(existing, neww) -> neww`, but it might also be nice to have like an `UpsertTransformer.OVERWRITE` singleton.